contrib/platform/src/com/sun/jna/platform/win32/WinNT.java

```diff
@@ -3126,33 +3134,84 @@ public void read() {
     /**
      * Represents information about a NUMA node in a processor group.
      */
-    @FieldOrder({ "nodeNumber", "reserved", "groupMask" })
+    @FieldOrder({ "nodeNumber", "reserved", "groupCount", "groupMasks" })
     public static class NUMA_NODE_RELATIONSHIP extends SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX {
 
         /**
-         * Identifies the NUMA node. Valid values are {@code 0} to the highest
-         * NUMA node number inclusive. A non-NUMA multiprocessor system will
-         * report that all processors belong to one NUMA node.
+         * Identifies the NUMA node. Valid values are {@code 0} to the highest NUMA node
+         * number inclusive. A non-NUMA multiprocessor system will report that all
+         * processors belong to one NUMA node.
          */
         public int nodeNumber;
 
         /**
          * This member is reserved.
          */
-        public byte[] reserved = new byte[20];
+        public byte[] reserved = new byte[18];
+
+        /**
+         * The number of groups included in the GroupMasks array. This field was
+         * introduced in TBD Release Iron. On earlier versions, this value is always 0.
+         */
+        public short groupCount;
 
         /**
          * A {@link GROUP_AFFINITY} structure that specifies a group number and
-         * processor affinity within the group.
+         * processor affinity within the group. This member is only relevant if
+         * {@code groupCount} is 0.
          */
         public GROUP_AFFINITY groupMask;
 
+        /**
+         * An array of {@link GROUP_AFFINITY} structures that specifies a group number
+         * and processor affinity within the group. This member is only relevant if
+         * {@code groupCount} is 1 or greater.
+         */
+        public GROUP_AFFINITY[] groupMasks = new GROUP_AFFINITY[1];
+
         public NUMA_NODE_RELATIONSHIP() {
         }
 
         public NUMA_NODE_RELATIONSHIP(Pointer memory) {
             super(memory);
         }
+
+        /*
+         * The groupMasks array default size is 1. Resize only if larger. Since the
+         * groupMasks array always has at least one element we copy it to the groupMask
+         * field. This is needed for backwards compatibility in the case groupCount is 0
+         * but done all the time for convenience.
+         */
+        @Override
+        public void read() {
+            // resize array if needed
+            readField("groupCount");
+            if (groupCount > 1) {
+                groupMasks = new GROUP_AFFINITY[groupCount];
+            }
```

Comment:
I suggest to replace with:

```suggestion
            groupMasks = new GROUP_AFFINITY[Math.max(groupCount, 1)];
```

that way multiple reads will work, even if the structure is wrapped once over old style Structure and once over new style structure.

Reply:
Whether it's "old" (0) or "new" (>0) is an OS-level condition that won't change during execution.  But conceivably if someone wanted to re-use a structure and just change the native bits with `useMemory()` the value could change.  But in this case, the proposed change would recreate the array (and reinitializing/recreating its internal Structure objects and the `ULONG_PTR`) on every `read()`.  What we really want is to conditionally resize the array only if it's a new version and has changed.

So I'd propose:
```
if (groupCount > 0 && groupCount != groupMaks.length) {
    groupMasks = new GROUP_AFFINITY[groupCount];
}
```

Reply:
Looking at this again, while I think it would be highly unusual for this value to change on re-read (requires both changing the number of processor groups while the program was operating (e.g., using a hypervisor and changing from under to over 64 logical processors) and re-using an existing structure rather than generating a new one) we do something similar to your proposal in the `PROCESSOR_RELATIONSHIP` structure, so it's a reasonable approach from the standpoint of consistency.

In the above edge case, the other structures in

Reply:
If you really want to optimize, you could do this (there needs to be at least one entry in the array (the original single value), reinitialization of the array is only needed if size changes:

```java
        @Override
        public void read() {
            // resize array if needed
            readField("groupCount");
            int requiredArraySize = Math.max(1, groupCount);
            if (groupMasks == null || requiredArraySize != groupMasks.length) {
                groupMasks = new GROUP_AFFINITY[requiredArraySize];
            }
            // Now read the fields from field order
            super.read();
            // Finally copy first value from array to older version of structure
            groupMask = groupMasks[0];
        }
```

Whether that is more efficient, that simple rebuilding the array from scratch and letting the GC do the work is a different discussion. The GC might win here.

Reply:
I don't think `groupMasks == null` will ever happen. We need a minimum array size of 1 and initialize it that way. Reading only needs to resize it if it needs to grow. (And maybe shrink on re-read.)

Reply:
OK, I'm done tweaking now.